RFCs/FS-1090-Generic-struct-type-whose-fields-are-all-unmanaged-types-is-unmanaged.md

```diff
@@ -0,0 +1,198 @@
+# F# RFC FS-1088 - (Generic struct type whose fields are all unmanaged types is unmanaged)
+
+The design suggestion [Generic struct type whose fields are all unmanaged types is unmanaged](https://github.com/fsharp/fslang-suggestions/issues/692) has been marked "approved in principle".
+This RFC covers the detailed proposal for this suggestion.
+
+* [x] Approved in principle
+* [x] [Suggestion](https://github.com/fsharp/fslang-suggestions/issues/692)
+* [ ] Implementation: [In progress](https://github.com/dotnet/fsharp/pull/6064)
+
+# Summary
+
+[summary]: #summary
+
+Allow generic structs to be `unmanaged` at construction if all fields are `unmanaged` types.
+
+# Motivation
+
+[motivation]: #motivation
+
+* Make it easier to author low level interop code.
+* Improve interoperability with C#.
+
+# Detailed design
+
+[design]: #detailed-design
+
+Currently, it is not possible to have an `unmanaged` constraint working on generic struct types.
+
+Consider following code:
+
+```fsharp
+[<Struct>]
+type Test<'T when 'T: unmanaged> =
+        val element: 'T
+
+let works = Test<int>()
+let error = Test<Test<int>>()
+
+```
+
+Prior to this RFC, the example above will fail to compile with:
+
+```less
+let error = Test<Test<int>>()
+  ------------^^^^^^^^^^^^^^^
+
+stdin(6,13): error FS0001: A generic construct requires that the type 'Test<int>' is an unmanaged type
+```
+
+This proposal aims to eliminate this restriction by treating generic struct type `unmanaged` if all its fields are `unmanaged`.
+
+## Definition
+
+An unmanaged-type is any type that isn't a reference-type and contains no fields whose type is not an unmanaged-type.
+
+In other words, an unmanaged-type is one of the following:
+
+* `sbyte, byte, short, ushort, int, uint, long, ulong, char, float, double, decimal, or bool`.
+* Any enum-type.
+* Any pointer-type.
+* Any generic user-defined struct-type that can be statically determined to be 'unmanaged' at construction.
+
+## Implementation details
```

Comment:
Needs some more info:

* Can I mix `unamnaged` with other generic constraints? Right now I can actually write the following code that looks like absolute nonsense.

```fsharp
[<Struct>]
type Test<'T when 'T: unmanaged and 'T: not struct> =
    val element: 'T
```

This and other silly things you can do today should be corrected as a part of this change. `unmanaged` implies `struct` in C#, which I would expect to be the case here. We'll need an example of doing this and what the error message should say.

* Clarify that the constraint is enforced by the language, not the runtime

* Clarify that nested fields in structs must also be unmanaged, with examples of working/not-working

Reply:
> * Can I mix `unamnaged` with other generic constraints? Right now I can actually write the following code that looks like absolute nonsense.

That is actually one of the things I wanted to add, and didn't, because I assumed it works already (I thought that you cannot mix 'unmanaged' and 'not struct' now).
I will add a point for fixing this.
Thanks for noticing.

Reply:
Added a point to implementation details.